### PR TITLE
[REVIEW] DataFrame `insert` and creation optimizations

### PR DESCRIPTION
**Pull request by galipremsagar**
This PR removes double index equality & reindexing checks in `DataFrame` construction and `insert` code-flows.

`scalar_broadcast_to`, a **2.2x speedup** for `str` dtype, numeric types perf remains the same:
```python

# This PR:

In [12]: %timeit cudf.utils.utils.scalar_broadcast_to('abc', 400000000, "str")
100 ms ± 208 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

# branch-22.04
In [25]: %timeit cudf.utils.utils.scalar_broadcast_to('abc', 400000000, "str")
227 ms ± 10 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

`assign`, a **1.2x-1.4x speedup**:
```python

In [9]: import numpy as np

In [10]: df = cudf.DataFrame(
    ...:     {
    ...:         "a": [1, 2, 3, np.nan] * 10000000,
    ...:         "b": ["a", "b", "c", "d"] * 10000000,
    ...:         "c": [0.0, 0.12, np.nan, 10.12] * 10000000,
    ...:     },
    ...: )

# THIS PR
In [11]: %timeit df.assign(f=10)
28.1 ms ± 365 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

# branch-22.04
In [4]: %timeit df.assign(f=10)
35 ms ± 10.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)



# THIS PR
In [4]: %timeit df.assign(g='hello world')
38.4 ms ± 250 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

# branch-22.04
In [4]: %timeit df.assign(g='hello world')
54 ms ± 82.6 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

DataFrame constructor, a **2x speedup**:

```python

In [2]: s1 = cudf.Series([1, 2, 3] * 1000000)

In [4]: s1.index = s1.index.astype("float64")

In [5]: s2 = cudf.Series([2, 3, 3] * 1000000)

In [6]: s2.index = s2.index.astype("float64")


In [12]: s3 = cudf.Series([10, 11, 12] * 1000000)

In [13]: s3.index = s3.index.astype("float64")

# THIS PR
In [14]: %timeit cudf.DataFrame({'a':s1, 'b':s2, 'c':s3})
2.81 ms ± 63.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

# branch-22.04
In [9]: %timeit cudf.DataFrame({'a':s1, 'b':s2, 'c':s3})
5.37 ms ± 188 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

### CODE REVIEW

**Comment by galipremsagar** [comment]
cc: @randerzander @ayushdg

File: python/cudf/cudf/core/dataframe.py
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:2567]: It looks like the main optimization here is that indexes should not be ignored by the public API, but you've found internal usage where we should. You mentioned trying to apply #10263 to `insert` once that's merged, so I think the default parameter here should maybe be `ignore_index=True` so that we only use the index when needed. That way switching APIs will lead to the fast behavior by default (assuming that test will easily catch cases where the code actually needs to account for the index), otherwise it's not enough to switch APIs but users would also need to specify the right parameter value.

File: python/cudf/cudf/core/dataframe.py
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:2557]: Unrelated, but I notice that `nan_as_null` is not documented and it should be.

File: python/cudf/cudf/core/dataframe.py
Comment (galipremsagar) [reply python/cudf/cudf/core/dataframe.py]: Sure, `ignore_index` should `True` by default. Changed it 👍

File: python/cudf/cudf/core/dataframe.py
Comment (galipremsagar) [reply python/cudf/cudf/core/dataframe.py]: Added docs

File: python/cudf/cudf/core/dataframe.py
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:839]:
```suggestion
                    nan_as_null=nan_as_null,
```

File: python/cudf/cudf/core/dataframe.py
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:2614]: There's probably a faster way to construct this if we know the input is a `cudf.Series`. I'm not sure how much we could save for a pandas Series by handling it manually, I don't think we typically do anything special for those. That may be worth exploring in a future PR (basically seeing if we can implement `Series.from_pandas` in a more efficient manner than just calling the constructor), but is out of scope for now.

File: python/cudf/cudf/core/dataframe.py
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:4747]:
```suggestion
                        self.index._data[col_name],
```

File: python/cudf/cudf/core/dataframe.py
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:5740]:
```suggestion
                df._insert(len(df._data), k, col)
```

File: python/cudf/cudf/core/dataframe.py
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:6551]:
```suggestion
                    value=replace_df[col_2],
```

File: python/cudf/cudf/core/frame.py
Comment (vyasr) [inline python/cudf/cudf/core/frame.py:6813]:
```suggestion
            value=join_res._index,
```

**Review by vyasr — APPROVED** [review]
LGTM. Mostly just suggestions to remove the default parameter since we changed it and that should be the default expected behavior for `_insert`. Other than that, one suggestion where you might be able to optimize a constructor. I'll leave it to you to finalize.

File: python/cudf/cudf/core/dataframe.py
Comment (galipremsagar) [reply python/cudf/cudf/core/dataframe.py:2614]: Yeah, I think it is out of scope for this PR. The reason being we have other places in the code-base which use the similar patterns we might be better off tackling that in a separate PR.

File: python/cudf/cudf/core/dataframe.py
Comment (galipremsagar) [reply python/cudf/cudf/core/dataframe.py]: Somehow I can't add suggestion to batch or commit your suggestion in this review 😕 so pushed separate commit.

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**
